include/aws/sdkutils/private/endpoints_ruleset_types_impl.h

```diff
@@ -0,0 +1,159 @@
+    struct aws_string *documentation;
+    bool is_deprecated;
+    struct aws_string *deprecated_message;
+    struct aws_string *deprecated_since;
+};
+
+struct aws_endpoints_ruleset {
+    struct aws_allocator *allocator;
+    struct aws_ref_count ref_count;
+
+    /* list of (aws_endpoints_rule) */
+    struct aws_array_list rules;
+
+    struct aws_string *version;
+    struct aws_string *service_id;
+    /* map of (aws_byte_cursor *) -> (aws_endpoints_parameter *) */
+    struct aws_hash_table *parameters;
+};
+
+struct aws_endpoints_function {
+    struct aws_string *fn; /* TODO: precompute hash or enum value to avoid lookup on every eval */
+    /* List of (aws_endpoints_expr) */
+    struct aws_array_list argv;
+};
+
+struct aws_endpoints_expr {
+    enum aws_endpoints_expr_type type;
+    union {
+        struct aws_string *string;
+        double number;
+        bool boolean;
+        struct aws_array_list array; /* List of (aws_endpoints_expr) */
+        struct aws_string *reference;
+        struct aws_endpoints_function function;
+    } e;
+};
+
+struct aws_endpoints_rule_data_endpoint {
+    struct aws_allocator *allocator;
+    enum aws_endpoints_url_type url_type;
+    union {
+        struct aws_string *template;
+        struct aws_string *reference;
+        struct aws_endpoints_function function;
+    } url;
+
+    /*
+     * Note: this is a custom properties json associated with the result.
+     * Properties are unstable and format can change frequently.
+     * Its up to caller to parse json to retrieve properties.
+     */
+    struct aws_string *properties;
+    /* Map of (aws_string *) -> (aws_array_list * of aws_string *) */
+    struct aws_hash_table *headers;
+};
+
+struct aws_endpoints_rule_data_error {
+    enum aws_endpoints_error_type error_type;
+    union {
+        struct aws_string *template;
+        struct aws_string *reference;
+        struct aws_endpoints_function function;
+    } error;
+};
+
+struct aws_endpoints_rule_data_tree {
+    /* List of (aws_endpoints_rule) */
+    struct aws_array_list rules;
+};
+
+struct aws_endpoints_condition {
+    struct aws_endpoints_function function;
+    struct aws_string *assign;
+};
+
+struct aws_endpoints_rule {
+    struct aws_allocator *allocator;
+
+    /* List of (aws_endpoints_condition) */
+    struct aws_array_list conditions;
+    struct aws_string *documentation;
+
+    enum aws_endpoints_rule_type type;
+    union {
+        struct aws_endpoints_rule_data_endpoint endpoint;
+        struct aws_endpoints_rule_data_error error;
+        struct aws_endpoints_rule_data_tree tree;
+    } rule_data;
+};
+
+struct aws_endpoints_parameter *aws_endpoints_parameter_new(
+    struct aws_allocator *allocator,
+    enum aws_endpoints_parameter_value_type type,
+    const struct aws_byte_cursor *name_cur);
+void aws_endpoints_parameter_destroy(struct aws_endpoints_parameter *parameter);
+
+void aws_endpoints_rule_cleanup(struct aws_endpoints_rule *rule);
```

Comment:
utterly trivial: in our existing codebase it's pretty much always `_clean_up()` not ~_cleanup()~